src/context/websocketContext.jsx

```diff
@@ -54,20 +55,62 @@ export const SocketProvider = (props) => {
 
   PubSub.setOnSubscriptionsChange((newTopics) => updateTopicsDebounce(newTopics))
 
-  const onMessage = (message) => {
-    const data = JSON.parse(message.data)
-    if (data.topic === 'heartbeat') return
+  const [overloaded, setOverloaded] = useState(false)
+  const [toastShown, setToastShown] = useState(false)
 
-    if (data.topic === 'server.restart_requested') setServerRestartingVisible(true)
+  // when overloaded is true, activate toast
+  useEffect(() => {
+    if (overloaded)
+      if (!toastShown) {
+        toast.warning(<RefreshToast />, {
+          autoClose: false,
+          closeButton: false,
+        })
+        setToastShown(true)
+      }
 
-    if (data.sender === window.senderId) {
-      return // my own message. ignore
+    return () => {
+      setOverloaded(false)
     }
-    if (data.topic === 'shout' && data?.summary?.text) toast.info(data.summary.text)
+  }, [overloaded, setOverloaded])
 
-    console.log('Event RX', data)
-    PubSub.publish(data.topic, data)
-  }
+  // onMessage is a function that is called when a message comes in from the websocket
+  // it is a closure that keeps track of the number of calls and the last call time
+  let onMessage = (() => {
+    let callCount = 0
+    let lastCall = Date.now()
+
+    return (message) => {
+      // If the function is called more than 10 times per second, return early.
+      if (callCount > 100) {
+        setOverloaded(true)
+        return console.log('WS OVERLOAD!!!!')
+      }
+
+      const data = JSON.parse(message.data)
```

Comment:
What do you think about destructuring ? Say you would write something like 
`const { topic, sender, summary } = data;`

So you do not have to extract it from data everywhere. You would get something like 

```
  if (topic === 'heartbeat') return;
  if (topic === 'server.restart_requested') {
    setServerRestartingVisible(true);
    return;
  }
  if (sender === window.senderId) return;
```
Not pushing, just asking.